File: .dev-tools/extract-writing-images.js
const fs = require("fs");
const path = require("path");
const unzipper = require("unzipper");

const WORD_DIR = "H:\\IT\\单词化句\\语料库\\word版英语真题";
const IMG_DIR = "H:\\IT\\单词化句\\vocab-reading-generator\\images\\writing";
const MOBILE_IMG_DIR = "H:\\IT\\单词化句\\vocab-reading-generator-mobile\\images\\writing";
const WWW_IMG_DIR = "H:\\IT\\单词化句\\vocab-reading-generator-mobile\\www\\images\\writing";

const YEAR_MAP = {
    "1999": "1999-1", "2000": "2000-1", "2001": "2001-1", "2002": "2002-1",
    "2003": "2003-1", "2004": "2004-1", "2005": "2005-1", "2006": "2006-1",
    "2007": "2007-1", "2008": "2008-1", "2009": "2009-1",
    "2010-1": "2010-1-1", "2010-2": "2010-2-2",
    "2011-1": "2011-1-1", "2011-2": "2011-2-2",
    "2012-1": "2012-1-1", "2012-2": "2012-2-2",
    "2013-1": "2013-1-1", "2013-2": "2013-2-2",
    "2014-1": "2014-1-1", "2014-2": "2014-2-2",
    "2015-1": "2015-1-1", "2015-2": "2015-2-2",
    "2016-1": "2016-1-1", "2016-2": "2016-2-2",
    "2017-1": "2017-1-1", "2017-2": "2017-2-2",
    "2018-1": "2018-1-1", "2018-2": "2018-2-2",
    "2019-1": "2019-1-1", "2019-2": "2019-2-2",
    "2020-1": "2020-1-1", "2020-2": "2020-2-2",
    "2021-1": "2021-1-1", "2021-2": "2021-2-2",
    "2022-1": "2022-1-1", "2022-2": "2022-2-2",
    "2023-1": "2023-1-1", "2023-2": "2023-2-2",
    "2024-1": "2024-1-1", "2024-2": "2024-2-2",
    "2025-1": "2025-1-1", "2025-2": "2025-2-2"
};

async function extractImagesFromDocx(docxPath, prefix) {
    const images = [];
    return new Promise((resolve, reject) => {
        fs.createReadStream(docxPath)
            .pipe(unzipper.Parse())
            .on("entry", function(entry) {
                const fileName = entry.path;
                if (fileName.startsWith("word/media/") &&
                    (fileName.endsWith(".png") || fileName.endsWith(".jpeg") ||
                     fileName.endsWith(".jpg") || fileName.endsWith(".gif"))) {
                    const ext = path.extname(fileName).substring(1);
                    const imgIndex = images.length;
                    const suffix = imgIndex === 0 ? "" : "-" + imgIndex;
                    const outName = prefix + "-partB" + suffix + "." + ext;
                    entry.buffer().then(buffer => {
                        images.push({ fileName: outName, buffer: buffer });
                    });
                } else {
                    entry.autodrain();
                }
            })
            .on("close", () => resolve(images))
            .on("error", reject);
    });
}

function saveImage(imageData, targetDir) {
    const filePath = path.join(targetDir, imageData.fileName);
    if (!fs.existsSync(targetDir)) {
        fs.mkdirSync(targetDir, { recursive: true });
    }
    fs.writeFileSync(filePath, imageData.buffer);
    return filePath;
}

function imageExists(prefix, targetDir) {
    const files = fs.existsSync(targetDir) ? fs.readdirSync(targetDir) : [];
    return files.some(f => f.startsWith(prefix));
}

async function main() {
    console.log("开始从Word文档提取作文图片...\n");
    const docxFiles = fs.readdirSync(WORD_DIR).filter(f => f.endsWith(".docx"));
    let totalExtracted = 0;
    let skipped = 0;

    for (const docxFile of docxFiles) {
        const yearKey = docxFile.replace(".docx", "");
        const prefix = YEAR_MAP[yearKey];
        if (!prefix) {
            console.log("跳过未知年份: " + docxFile);
            continue;
        }
        if (imageExists(prefix, IMG_DIR)) {
            skipped++;
            continue;
        }
        console.log("提取中: " + yearKey + "...");
        try {
            const docxPath = path.join(WORD_DIR, docxFile);
            const images = await extractImagesFromDocx(docxPath, prefix);
            if (images.length > 0) {
                console.log("  找到 " + images.length + " 张图片");
                for (const img of images) {
                    saveImage(img, IMG_DIR);
                    saveImage(img, MOBILE_IMG_DIR);
                    saveImage(img, WWW_IMG_DIR);
                    console.log("    保存: " + img.fileName);
                    totalExtracted++;
                }
            } else {
                console.log("  未找到图片");
            }
        } catch (err) {
            console.log("  解析失败: " + err.message);
        }
    }
    console.log("\n完成！共提取 " + totalExtracted + " 张图片，跳过 " + skipped + " 个");
}

main().catch(console.error);

File: .dev-tools/extract-writing-images.py
import os
import re
from docx import Document
from docx.oxml.ns import qn

WORD_DIR = r'h:\IT\单词化句\语料库\word版英语真题'
IMG_DIR = r'h:\IT\单词化句\vocab-reading-generator\images\writing'


def get_paragraph_image_rids(paragraph):
    rids = []
    elem = paragraph._element
    xml_str = elem.xml
    embed_matches = re.findall(r'r:embed="([^"]+)"', xml_str)
    rids.extend(embed_matches)
    id_matches = re.findall(r'r:id="([^"]+)"', xml_str)
    rids.extend(id_matches)
    return list(set(rids))


def extract_partb_image(docx_path, paper_id):
    doc = Document(docx_path)
    
    part_b_start = -1
    for i, para in enumerate(doc.paragraphs):
        text = para.text.strip()
        if i < 30:
            continue
        if 'Part B' in text and ('Directions' in text or len(text) < 20):
            part_b_start = i
            break
    
    if part_b_start < 0:
        for i, para in enumerate(doc.paragraphs):
            text = para.text.strip()
            if 'Part B' in text and i > 50:
                part_b_start = i
                break
    
    if part_b_start < 0:
        return None
    
    rid_to_image = {}
    for rel in doc.part.rels.values():
        if 'image' in rel.reltype:
            image_part = rel.target_part
            img_bytes = image_part.blob
            ext = image_part.content_type.split('/')[-1]
            if ext == 'jpeg':
                ext = 'jpg'
            rid_to_image[rel.rId] = {
                'bytes': img_bytes,
                'ext': ext,
                'size': len(img_bytes)
            }
    
    part_b_images = []
    found_part_b = False
    
    for i in range(part_b_start, len(doc.paragraphs)):
        para = doc.paragraphs[i]
        text = para.text.strip()
        
        if 'Section III' in text or 'Section Ⅲ' in text:
            break
        if ('答案' in text or '参考答案' in text) and len(part_b_images) > 0:
            break
        
        img_rids = get_paragraph_image_rids(para)
        for rid in img_rids:
            if rid in rid_to_image and rid not in [x['rid'] for x in part_b_images]:
                img = rid_to_image[rid]
                img['rid'] = rid
                part_b_images.append(img)
    
    if len(part_b_images) == 0 and len(rid_to_image) > 0:
        max_size = 0
        max_img = None
        for rid, img in rid_to_image.items():
            if img['size'] > max_size:
                max_size = img['size']
                max_img = img
        if max_img and max_img['size'] > 5000:
            max_img['rid'] = 'fallback'
            part_b_images.append(max_img)
    
    part_b_images = [img for img in part_b_images if img['size'] > 1000]
    
    return part_b_images


def main():
    word_files = sorted([f for f in os.listdir(WORD_DIR) if f.endswith('.docx')])
    
    extracted = []
    skipped = []
    errors = []
    
    for fname in word_files:
        paper_id = fname.replace('.docx', '')
        docx_path = os.path.join(WORD_DIR, fname)
        
        try:
            images = extract_partb_image(docx_path, paper_id)
            
            if images is None or len(images) == 0:
                skipped.append(f'{paper_id}: PartB无图片')
                continue
            
            img = images[0]
            ext = img['ext']
            
            target_name = f'{paper_id}-partb.{ext}'
            target_path = os.path.join(IMG_DIR, target_name)
            
            old_files = [f for f in os.listdir(IMG_DIR) if f.startswith(f'{paper_id}-') and 'partb' in f.lower()]
            
            with open(target_path, 'wb') as f:
                f.write(img['bytes'])
            
            info = f'{paper_id}: 提取图片 -> {target_name} ({img["size"]} bytes)'
            if old_files:
                info += f' [原有: {", ".join(old_files)}]'
            extracted.append(info)
        
        except Exception as e:
            errors.append(f'{paper_id}: 错误 - {e}')
            import traceback
            traceback.print_exc()
    
    print('=== 已提取图片 ===')
    for s in extracted:
        print(s)
    
    print(f'\n总计：提取 {len(extracted)} 个，跳过 {len(skipped)} 个，错误 {len(errors)} 个')
    
    if errors:
        print('\n=== 错误 ===')
        for s in errors:
            print(s)


if __name__ == '__main__':
    main()
